File: 797.all-paths-from-source-to-target.java
/*
 * @lc app=leetcode id=797 lang=java
 *
 * [797] All Paths From Source to Target
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {
    private List<List<Integer>> results;
    private List<Integer> path;
    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        results = new ArrayList<>();
        path = new ArrayList<>();
        path.add(0);
        backtracking(graph, 0);
        return results;
    }
    private void backtracking(int[][] graph, int current) {
        if (path.get(path.size() - 1) == graph.length - 1) {
            results.add(new ArrayList<>(path));
            return;
        }
        if (path.size() == graph.length) return;
        for (int n : graph[current]) {
            path.add(n);
            backtracking(graph, n);
            path.removeLast();
        }
    }
}
// @lc code=end

File: 797.all-paths-from-source-to-target.cpp
/*
 * @lc app=leetcode id=797 lang=cpp
 *
 * [797] All Paths From Source to Target
 */

// @lc code=start

#include <vector>
#include <functional>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        vector<vector<int>> paths;
        vector<int> path; path.push_back(0);
        function<void(const int, const int)> dfs = [&](const int from, const int to) -> void {
            if (from == to) {
                paths.push_back(path); return;
            }
            for_each(graph[from].begin(), graph[from].end(), [&](const int n) {
                path.push_back(n);
                dfs(n, to);
                path.pop_back();
            });
        };
        dfs(0, graph.size() - 1);
        return paths;
    }
};
// @lc code=end
